### ISSUE TITLE
No type narrowing of subtypes in the match statement

### ISSUE THREAD

**Issue by williamw4096** [microsoft/pyright#7280]
**Describe the bug**
Pyright does not narrow the type for subsequent cases within a match statement to remove subclasses when a class pattern is used. I.e. if the expression could be of type B (a subtype of A), and the first match pattern matches any type A, Pyright will still think the expression could be a B in subsequent cases, even though this would have been caught by the A case, since class patterns are checked with isinstance().

**Code**
```python
from typing import assert_never

class A: pass
class B(A): pass

x = B()
match x:
    case A():
        pass
    case _ as unreachable:
        assert_never(unreachable)
```

Expected behaviour: No type error, since anything of type B should be caught by the first case, and x is of type B
Actual behaviour: Argument of type "B" cannot be assigned to parameter "__arg" of type "Never"

**VS Code extension or command-line**
Pyright command line 1.1.350

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.351, which I just published.
